pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/MPOTNCuda.hpp

```diff
@@ -0,0 +1,221 @@
+    }
+
+    /**
+     * @brief Get a vector of pointers to tensor data of each site.
+     *
+     * @return std::vector<void *>
+     */
+    [[nodiscard]] auto getTensorsDataPtr_() -> std::vector<void *> {
+        std::vector<void *> tensorsDataPtr;
+        for (auto &tensor : tensors_) {
+            tensorsDataPtr.emplace_back(
+                reinterpret_cast<void *>(tensor->getDataBuffer().getData()));
+        }
+        return tensorsDataPtr;
+    }
+
+  public:
+    explicit MPOTNCuda(const std::vector<std::vector<ComplexT>> &tensors,
+                       const std::vector<std::size_t> &wires,
+                       const std::size_t maxBondDim,
+                       const std::size_t numQubits,
+                       const cutensornetHandle_t &cutensornetHandle,
+                       const cudaDataType_t &cudaDataType,
+                       const DevTag<int> &dev_tag) {
+        PL_ABORT_IF_NOT(tensors.size() == wires.size(),
+                        "Number of tensors and wires must match.");
+
+        PL_ABORT_IF(maxBondDim < 2,
+                    "Max MPO bond dimension must be at least 2.");
+
+        PL_ABORT_IF_NOT(std::is_sorted(wires.begin(), wires.end()),
+                        "Only sorted target wires is accepeted.");
+
+        PL_ABORT_IF_NOT(wires.size() == wires.back() - wires.front() + 1,
+                        "Only support local target wires.");
+
+        // Create an empty MPO tensor network operator. Note that the state
+        // extents are aligned with the quantum state.
+        PL_CUTENSORNET_IS_SUCCESS(cutensornetCreateNetworkOperator(
+            /* const cutensornetHandle_t */ cutensornetHandle,
+            /* int32_t */ static_cast<int32_t>(numQubits),
+            /* const int64_t stateModeExtents */
+            std::vector<int64_t>(numQubits, 2).data(),
+            /* cudaDataType_t */ cudaDataType,
+            /* cutensornetNetworkOperator_t */ &MPOOperator_));
+
+        numMPOSites_ = wires.size();
+
+        MPO_modes_int32_.resize(numMPOSites_);
+
+        std::iota(MPO_modes_int32_.begin(), MPO_modes_int32_.end(),
+                  wires.front());
+
+        std::transform(MPO_modes_int32_.begin(), MPO_modes_int32_.end(),
+                       MPO_modes_int32_.begin(),
+                       [&numQubits](const std::size_t mode) {
+                           return static_cast<int32_t>(numQubits - 1 - mode);
+                       });
+
+        // Ensure the modes are in ascending order
+        std::reverse(MPO_modes_int32_.begin(), MPO_modes_int32_.end());
+
+        for (std::size_t i = 0; i < numMPOSites_ - 1; i++) {
+            std::size_t bondDim = std::min(i + 1, numMPOSites_ - i - 1) *
+                                  2; // 1+1 (1 for bra and 1 for ket)
+            if (bondDim <= log2(maxBondDim)) {
+                bondDims_.emplace_back(std::size_t{1} << bondDim);
+            } else {
+                bondDims_.emplace_back(maxBondDim);
```

Comment:
I am a bit concerned about using `emplace_back` in too many parts of the code. Do you think exploring an alternative to creating these arrays as a not consecutive list, for example, a linked list and then collapsing the construction in a sequential list such as a `vector`